agent/doctor/doctor.go

```diff
@@ -35,72 +30,92 @@ type Doctor struct {
 	lock                 sync.RWMutex
 	cluster              string
 	containerInstanceArn string
+	statusReported       bool
 }
 
 func NewDoctor(healthchecks []Healthcheck, cluster string, containerInstanceArn string) (*Doctor, error) {
 	newDoctor := &Doctor{
 		healthchecks:         []Healthcheck{},
 		cluster:              cluster,
 		containerInstanceArn: containerInstanceArn,
+		statusReported:       false,
 	}
 	for _, hc := range healthchecks {
 		newDoctor.AddHealthcheck(hc)
 	}
 	return newDoctor, nil
 }
 
+// GetCluster returns the cluster that was provided to the doctor while
+// being initialized
+func (doc *Doctor) GetCluster() string {
+	doc.lock.RLock()
+	defer doc.lock.RUnlock()
+
+	return doc.cluster
+}
+
+// GetContainerInstanceArn returns the container instance arn that was
+// provided to the doctor while being initialized
+func (doc *Doctor) GetContainerInstanceArn() string {
+	doc.lock.RLock()
+	defer doc.lock.RUnlock()
+
+	return doc.containerInstanceArn
+}
+
+// SetStatusReported tells the doctor that we have already reported the
+// current status of the healthchecks to the backend
+func (doc *Doctor) SetStatusReported(statusReported bool) {
+	doc.lock.Lock()
+	defer doc.lock.Unlock()
+
+	doc.statusReported = statusReported
+}
+
+// HasStatusBeenReported returns whether we have already sent the current
+// state of the healthchecks to the backend or not
+func (doc *Doctor) HasStatusBeenReported() bool {
+	doc.lock.RLock()
+	defer doc.lock.RUnlock()
+
+	return doc.statusReported
+}
+
+// AddHealthcheck adds a healthcheck to the list of healthchecks that the
+// doctor will run every time doctor.RunHealthchecks() is called
 func (doc *Doctor) AddHealthcheck(healthcheck Healthcheck) {
 	doc.lock.Lock()
 	defer doc.lock.Unlock()
 	doc.healthchecks = append(doc.healthchecks, healthcheck)
 }
 
+// RunHealthchecks runs every healthcheck that the doctor knows about and
+// returns a cumulative result; true if they all pass, false otherwise
 func (doc *Doctor) RunHealthchecks() bool {
-	doc.lock.RLock()
-	defer doc.lock.RUnlock()
+	doc.lock.Lock()
+	defer doc.lock.Unlock()
 	allChecksResult := []HealthcheckStatus{}
 
 	for _, healthcheck := range doc.healthchecks {
 		res := healthcheck.RunCheck()
 		seelog.Debugf("instance healthcheck result: %v", res)
 		allChecksResult = append(allChecksResult, res)
 	}
+
+	doc.statusReported = false
```

Comment:
Set it using `SetStatusReported()` so that it's under a lock ?

Reply:
We are under a lock already. If I call `SetStatusReported()`, we would run into a deadlock because `SetStatusReported()` would be waiting for a lock that `RunHealthchecks()` is holding and `RunHealthchecks()` won't release the lock until the end of the method.